File: app/services/media_safety.py
from __future__ import annotations

import asyncio
from io import BytesIO

from aiogram import Bot
from aiogram.types import Message


MAX_SLIDESHOW_IMAGE_BYTES = 20 * 1024 * 1024
MAX_SLIDESHOW_VIDEO_BYTES = 50 * 1024 * 1024
MAX_MEDIA_DIMENSION = 10_000
MAX_IMAGE_PIXELS = 40_000_000
TELEGRAM_DOWNLOAD_TIMEOUT = 15.0


class UnsafeMediaError(ValueError):
    pass


def _validate_dimensions(width: int | None, height: int | None) -> None:
    if not width or not height:
        return
    if width > MAX_MEDIA_DIMENSION or height > MAX_MEDIA_DIMENSION:
        raise UnsafeMediaError("media dimensions exceed the allowed limit")
    if width * height > MAX_IMAGE_PIXELS:
        raise UnsafeMediaError("media pixel count exceeds the allowed limit")


def validate_slideshow_message(message: Message) -> None:
    if message.photo:
        media = message.photo[-1]
        if media.file_size and media.file_size > MAX_SLIDESHOW_IMAGE_BYTES:
            raise UnsafeMediaError("photo is too large")
        _validate_dimensions(media.width, media.height)
        return
    if message.video:
        media = message.video
        if media.file_size and media.file_size > MAX_SLIDESHOW_VIDEO_BYTES:
            raise UnsafeMediaError("video is too large")
        _validate_dimensions(media.width, media.height)


async def safe_telegram_download(
    bot: Bot,
    file_id: str,
    *,
    max_bytes: int = MAX_SLIDESHOW_IMAGE_BYTES,
    timeout: float = TELEGRAM_DOWNLOAD_TIMEOUT,
) -> bytes:
    """Bound any Telegram file download by both time and bytes."""
    target = BytesIO()
    async with asyncio.timeout(timeout):
        await bot.download(file_id, destination=target)
    payload = target.getvalue()
    if len(payload) > max_bytes:
        raise UnsafeMediaError("downloaded Telegram file is too large")
    return payload


__all__ = [
    "MAX_IMAGE_PIXELS",
    "MAX_MEDIA_DIMENSION",
    "MAX_SLIDESHOW_IMAGE_BYTES",
    "MAX_SLIDESHOW_VIDEO_BYTES",
    "TELEGRAM_DOWNLOAD_TIMEOUT",
    "UnsafeMediaError",
    "safe_telegram_download",
    "validate_slideshow_message",
]

File: lib/media_safety.js
import { api } from 'sdk';
import { UnsafeMediaError } from 'lib/errors';
export const MAX_SLIDESHOW_IMAGE_BYTES=20*1024*1024;
export const MAX_SLIDESHOW_VIDEO_BYTES=50*1024*1024;
export const MAX_MEDIA_DIMENSION=10000;
export const MAX_IMAGE_PIXELS=40000000;
function validateDimensions(width,height){ if(!width||!height) return; if(Number(width)>MAX_MEDIA_DIMENSION||Number(height)>MAX_MEDIA_DIMENSION) throw new UnsafeMediaError('media dimensions exceed the allowed limit'); if(Number(width)*Number(height)>MAX_IMAGE_PIXELS) throw new UnsafeMediaError('media pixel count exceeds the allowed limit'); }
export function validateSlideshowMessage(message){ if(Array.isArray(message?.photo)&&message.photo.length){ const m=message.photo.at(-1); if(m.file_size&&Number(m.file_size)>MAX_SLIDESHOW_IMAGE_BYTES) throw new UnsafeMediaError('photo is too large'); validateDimensions(m.width,m.height); return; } if(message?.video){ const m=message.video; if(m.file_size&&Number(m.file_size)>MAX_SLIDESHOW_VIDEO_BYTES) throw new UnsafeMediaError('video is too large'); validateDimensions(m.width,m.height); } }
export async function safeTelegramDownload(fileId,{maxBytes=MAX_SLIDESHOW_IMAGE_BYTES}={}){ const payload=await api.getFileContent(fileId); let bytes; if(payload instanceof Uint8Array) bytes=payload; else if(payload instanceof ArrayBuffer) bytes=new Uint8Array(payload); else if(typeof payload==='string') bytes=new TextEncoder().encode(payload); else bytes=new Uint8Array(payload??[]); if(bytes.byteLength>maxBytes) throw new UnsafeMediaError('downloaded Telegram file is too large'); return bytes; }
